src/cache_cf.cc

```diff
@@ -1014,87 +1017,133 @@ parse_obsolete(const char *name)
 
-    *tptr = static_cast<time_msec_t>(m * d);
+    self_destruct();
+    return false;
+}
 
-    if (static_cast<double>(*tptr) * 2 != m * d * 2) {
-        debugs(3, DBG_CRITICAL, "FATAL: Invalid value '" <<
-               d << " " << (token ? token : units) << ": integer overflow (time_msec_t).");
+static void
+CheckTimeOverflow(const double value, const std::chrono::nanoseconds &unit) {
+    const auto maxNanoseconds = std::chrono::nanoseconds::max().count();
+    if (value > maxNanoseconds/static_cast<double>(unit.count())) {
+        const auto maxYears = maxNanoseconds/(HoursPerYear*3600*1000000000);
+        debugs(3, DBG_CRITICAL, "FATAL: The time must be less than " << maxYears << " years");
         self_destruct();
     }
 }
 
-static uint64_t
-parseTimeUnits(const char *unit, bool allowMsec)
+/// Parses a time specification from the config file and
+/// returns the time as a chrono duration object of 'TimeUnit' type.
+/// \param defaultUnitName will be used if the time unit is missing,
+/// it should passed only if the configuration paramter being parsed
+/// excpects more arguments after time unit.
+template <class TimeUnit>
+TimeUnit
+parseTimeLine(const char *defaultUnitName = nullptr)
 {
-    if (allowMsec && !strncasecmp(unit, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
-        return 1;
-
-    if (!strncasecmp(unit, T_SECOND_STR, strlen(T_SECOND_STR)))
-        return 1000;
+    char *token = ConfigParser::NextToken();;
+    if (!token) {
+        self_destruct();
+        return TimeUnit::zero();
+    }
 
-    if (!strncasecmp(unit, T_MINUTE_STR, strlen(T_MINUTE_STR)))
-        return 60 * 1000;
+    const auto parsedValue = xatof(token);
 
-    if (!strncasecmp(unit, T_HOUR_STR, strlen(T_HOUR_STR)))
-        return 3600 * 1000;
+    std::chrono::nanoseconds parsedUnitDuration;
 
-    if (!strncasecmp(unit, T_DAY_STR, strlen(T_DAY_STR)))
-        return 86400 * 1000;
+    token = parsedValue ? ConfigParser::PeekAtToken() : nullptr;
 
-    if (!strncasecmp(unit, T_WEEK_STR, strlen(T_WEEK_STR)))
-        return 86400 * 7 * 1000;
+    if (parsedValue && parseTimeUnits<TimeUnit>(token, defaultUnitName, parsedUnitDuration))
+        (void)ConfigParser::NextToken();
 
-    if (!strncasecmp(unit, T_FORTNIGHT_STR, strlen(T_FORTNIGHT_STR)))
-        return 86400 * 14 * 1000;
+    CheckTimeOverflow(parsedValue, parsedUnitDuration);
 
-    if (!strncasecmp(unit, T_MONTH_STR, strlen(T_MONTH_STR)))
-        return static_cast<uint64_t>(86400) * 30 * 1000;
+    const std::chrono::nanoseconds resultDuration(static_cast<std::chrono::nanoseconds::rep>(parsedUnitDuration.count() * parsedValue));
 
-    if (!strncasecmp(unit, T_YEAR_STR, strlen(T_YEAR_STR)))
-        return static_cast<uint64_t>(86400 * 1000 * 365.2522);
+    // validate precisions (time-unit-small only)
+    if (TimeUnit(1) <= std::chrono::microseconds(1) && resultDuration.count() && resultDuration.count() <= 3) {
```

Comment:
The logic behind this condition is unclear. It needs a clarifying comment or perhaps named constants for obscure expressions like resultDuration.count(). _You_ know what this means. The reader will have a hard time guessing.

Reply:
It is still difficult to grasp what "result duration" means, especially when there is also "result", which is, presumably, is also some kind of duration. I suspect you can clarify things quite a bit if you rename `resultDuration` to `nanoseconds`.

```suggestion
        if (0 < nanoseconds.count() && nanoseconds.count() < 3) {
```

The `result` variable can stay as is because the code should not care/know about the actual time unit of that value. The `!result.count()` expression further down the code would still look problematic/mysterious, but perhaps that can be addressed like this:

```
if (nanoseconds.count() && !result.count()) { // underflow
    debugs(3, DBG_CRITICAL, "FATAL: Configured time is too small for its context (becomes zero after required internal conversions): " << parsedValue << ' ' << unitName);
```

> The level 0 warning message itself explains another condition

It tries to do that but fails IMO because the code the reader sees clearly shows that we can store/manipulate small values such as 2 nanoseconds. The problem here is that the warning is about what would happen in some _other_, invisible here, code. Squid is not measuring the entered value. It will be measuring something else (e.g., code execution delays), and it is _those_ measurements that are unlikely to be done with a nanosecond precision. I suggest using something like this:

```
debugs(..., YYY << "WARNING: Squid time measurement precision is likely to be far worse than the nanosecond-level precision implied by the configured value: " << parsedValue << ' ' << unitName);
```

Where `YYY` stands for whatever code we use to report the current parsing file:lineno location. IIRC, there is already code printing such location in other Squid messages related to parsing problems.

Reply:
Done at 87222b.